Foundation/String/UTF8.hs

```diff
@@ -1246,37 +1252,80 @@ readInteger :: String -> Maybe Integer
 readInteger str
     | sz == 0  = Nothing
     | otherwise =
-        let (# c, nextOfs #) = next str 0
-         in case c of
-            '-'           -> fmap negate (loop 0 nextOfs)
-            _ | isDigit c -> loop (fromDigit c) nextOfs
-              | otherwise -> Nothing
+         let (# modF, startOfs #) = case nextAscii str 0 of
+                                        -- '-'
+                                        (# 0x2d, True #) -> (# negate , 1 #)
+                                        _                -> (# id, 0 #)
+          in case decimalDigits 0 str startOfs of
+                (# acc, True, endOfs #) | endOfs > startOfs -> Just $ modF acc
+                _                                           -> Nothing
   where
     !sz = size str
-    loop :: Integer -> Offset Word8 -> Maybe Integer
-    loop !acc !ofs
-        | ofs .==# sz = Just acc
-        | otherwise   =
-            let (# c, nextOfs #) = next str ofs
-             in if isDigit c
-                    then loop (acc * 10 + fromDigit c) nextOfs
-                    else Nothing
-    isDigit c = c >= '0' && c <= '9'
-    fromDigit c = integralUpsize (c - '0')
 
 readNatural :: String -> Maybe Natural
 readNatural str
     | sz == 0  = Nothing
-    | otherwise = loop 0 0
+    | otherwise =
+        case decimalDigits 0 str 0 of
+            (# acc, True, endOfs #) | endOfs > 0 -> Just $ acc
+            _                                    -> Nothing
+  where
+    !sz = size str
+
+readFloatingExact :: String -> Maybe (Integer, Word, Natural)
```

Comment:
I'd recommend a bunch of comments on this. I've no idea what the Word represents, and no real idea how/what would be produced. I think these things work great with doc comments that are also tested.